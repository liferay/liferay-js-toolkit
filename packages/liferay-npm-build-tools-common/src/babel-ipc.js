import {resolve} from 'path';

global._babel_ipc_ = global._babel_ipc_ || {};

/**
 * Get an IPC value for a given Babel plugin
 * @param  {Object} state babel plugin state argument
 * @param  {Object|function} defaultValue the default value or a factory to get it if value is not set
 * @return {Object} the IPC value
 */
export function get(state, defaultValue) {
	let defaultValueFactory = defaultValue;

	if (typeof defaultValue !== 'function') {
		defaultValueFactory = () => defaultValue;
	}

	if (
		!state ||
		!state.file ||
		!state.file.opts ||
		!state.file.opts.filenameRelative
	) {
		return defaultValueFactory();
	}

	const key = resolve(state.file.opts.filenameRelative);

	if (!global._babel_ipc_.hasOwnProperty(key)) {
		return defaultValueFactory();
	}

	return global._babel_ipc_[key];
}

/**
 * Set an IPC value for a given file path.
 * @param  {String} filePath the path of the file being processed
 * @param  {Object} value the IPC value to set
 * @return {void}
 */
export function set(filePath, value) {
	global._babel_ipc_[resolve(filePath)] = value;
}

/**
 * Clear an IPC value for a given file path.
 * @param  {String} filePath the path of the file being processed
 * @return {void}
 */
export function clear(filePath) {
	delete global._babel_ipc_[resolve(filePath)];
}
