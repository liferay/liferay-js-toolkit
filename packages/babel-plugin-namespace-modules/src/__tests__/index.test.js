import * as babel from 'babel-core';
import * as babelIpc from 'liferay-npm-build-tools-common/lib/babel-ipc';
import PluginLogger from 'liferay-npm-build-tools-common/lib/plugin-logger';
import plugin from '../index';

describe('when called from babel', () => {
	it('correctly namespaces require modules', () => {
		const source = `
			define(function(){
				require('a-module');
				require('./a-local-module');
				require('fs');
			})
		`;

		const {code} = babel.transform(source, {
			filenameRelative: __filename,
			plugins: [plugin],
		});

		expect(code).toMatchSnapshot();
	});

	it('correctly namespaces define() dependencies', () => {
		const source = `
			define(['a-module', './a-local-module', 'fs'], function(){
			})
		`;

		const {code} = babel.transform(source, {
			filenameRelative: __filename,
			plugins: [plugin],
		});

		expect(code).toMatchSnapshot();
	});

	it('does not namespaces define() module name', () => {
		const source = `
			define('a-module', function(){
			})
		`;

		const {code} = babel.transform(source, {
			filenameRelative: __filename,
			plugins: [plugin],
		});

		expect(code).toMatchSnapshot();
	});

	it('correctly namespaces all together', () => {
		const source = `
			define('a-module', ['a-module', './a-local-module', 'fs'], function(){
				require('a-module');
				require('./a-local-module');
				require('fs');
			})
		`;

		const {code} = babel.transform(source, {
			filenameRelative: __filename,
			plugins: [plugin],
		});

		expect(code).toMatchSnapshot();
	});
});

describe('when called from liferay-npm-bundler', () => {
	let logger;

	beforeEach(() => {
		PluginLogger.set(__filename, (logger = new PluginLogger()));
		babelIpc.set(__filename, {rootPkgJson: require('./package.json')});
	});

	afterEach(() => {
		PluginLogger.delete(__filename);
	});

	it('correctly namespaces require modules', () => {
		const source = `
			define(function(){
				require('a-module');
				require('./a-local-module');
				require('fs');
			})
		`;

		const {code} = babel.transform(source, {
			filenameRelative: __filename,
			plugins: [plugin],
		});

		expect(code).toMatchSnapshot();
	});

	it('correctly namespaces define() dependencies', () => {
		const source = `
			define(['a-module', './a-local-module', 'fs'], function(){
			})
		`;

		const {code} = babel.transform(source, {
			filenameRelative: __filename,
			plugins: [plugin],
		});

		expect(code).toMatchSnapshot();
	});

	it('correctly namespaces define() module name', () => {
		const source = `
			define('a-module', function(){
			})
		`;

		const {code} = babel.transform(source, {
			filenameRelative: __filename,
			plugins: [plugin],
		});

		expect(code).toMatchSnapshot();
	});

	it('correctly namespaces all together', () => {
		const source = `
			define('a-module', ['a-module', './a-local-module', 'fs'], function(){
				require('a-module');
				require('./a-local-module');
				require('fs');
			})
		`;

		const {code} = babel.transform(source, {
			filenameRelative: __filename,
			plugins: [plugin],
		});

		expect(code).toMatchSnapshot();
	});

	it('logs results correctly', () => {
		const source = `
			define('a-module', ['a-module', './a-local-module', 'fs'], function(){
				require('a-module');
				require('./a-local-module');
				require('fs');
			})
		`;

		babel.transform(source, {
			filenameRelative: __filename,
			plugins: [plugin],
		});

		expect(logger.messages).toMatchSnapshot();
	});
});
