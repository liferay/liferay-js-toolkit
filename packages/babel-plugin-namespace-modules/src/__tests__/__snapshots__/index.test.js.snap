// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when called from babel correctly namespaces all together 1`] = `
"
define('a-module', ['test-package$1.0.0$a-module', './a-local-module', 'fs'], function () {
	require('test-package$1.0.0$a-module');
	require('./a-local-module');
	require('fs');
});"
`;

exports[`when called from babel correctly namespaces define() dependencies 1`] = `
"
define(['test-package$1.0.0$a-module', './a-local-module', 'fs'], function () {});"
`;

exports[`when called from babel correctly namespaces require modules 1`] = `
"
define(function () {
	require('test-package$1.0.0$a-module');
	require('./a-local-module');
	require('fs');
});"
`;

exports[`when called from babel does not namespaces define() module name 1`] = `
"
define('a-module', function () {});"
`;

exports[`when called from liferay-npm-bundler correctly namespaces all together 1`] = `
"
define('test-package$1.0.0$a-module', ['test-package$1.0.0$a-module', './a-local-module', 'fs'], function () {
	require('test-package$1.0.0$a-module');
	require('./a-local-module');
	require('fs');
});"
`;

exports[`when called from liferay-npm-bundler correctly namespaces define() dependencies 1`] = `
"
define(['test-package$1.0.0$a-module', './a-local-module', 'fs'], function () {});"
`;

exports[`when called from liferay-npm-bundler correctly namespaces define() module name 1`] = `
"
define('test-package$1.0.0$a-module', function () {});"
`;

exports[`when called from liferay-npm-bundler correctly namespaces require modules 1`] = `
"
define(function () {
	require('test-package$1.0.0$a-module');
	require('./a-local-module');
	require('fs');
});"
`;

exports[`when called from liferay-npm-bundler logs results correctly 1`] = `
Array [
  Object {
    "level": "info",
    "source": "namespace-modules",
    "things": Array [
      "Namespaced",
      1,
      "define() names,",
      1,
      "define() dependencies,",
      "and",
      1,
      "require() names",
    ],
  },
]
`;
